src/GF.jl

```diff
@@ -18,54 +18,43 @@
 
 using KeldyshED: EDCore
 using Keldysh
-import Keldysh: θ
 using LinearAlgebra: Diagonal, tr
 using Distributed
-@everywhere using SharedArrays
 
 export computegf
 
-"""
-  Compute Green's function on a given Keldysh contour at inverse temperature β
-
-  If 'β' is omitted, 'grid' must be defined on a time contour containing the
-  imaginary branch.
-
-  This method returns a vector of TimeGF objects, one element per a pair of
-  indices in `c_cdag_index_pairs`.
-"""
-function computegf(ed::EDCore,
-                   grid::TimeGrid,
-                   c_cdag_index_pairs::Vector{Tuple{IndicesType, IndicesType}},
-                   β = nothing)
-
-  if isnothing(β)
-    im_b = get_branch(grid.contour, imaginary_branch)
-    if isnothing(im_b)
-      throw(DomainError("Cannot extract inverse temperature " *
-                        "-- no imaginary branch on the supplied contour"))
-    end
-    β = length(im_b)
-  end
+function _computegf!(ed::EDCore,
+                     gf::AbstractTimeGF{T, scalar},
+                     c_indices::Vector{IndicesType},
+                     cdag_indices::Vector{IndicesType},
+                     β::Float64) where {T <: Number, scalar}
 
   en = energies(ed)
   ρ = density_matrix(ed, β)
 
-  data = SharedArray{ComplexF64, 3}(length(c_cdag_index_pairs),
-                                    length(grid),
-                                    length(grid),
-                                    pids = workers())
-  all_jobs = [((ind_n, indices), (n1, t1), (n2, t2))
-              for (ind_n, indices) in enumerate(c_cdag_index_pairs)
-              for (n1, t1) in enumerate(grid)
-              for (n2, t2) in enumerate(grid)]
+  norb = norbitals(gf)
+  D = TimeDomain(gf)
+
+  all_jobs = [((c_n, c_index), (cdag_n, cdag_index), (t1, t2))
+              for (c_n, c_index) in enumerate(c_indices)
+              for (cdag_n, cdag_index) in enumerate(cdag_indices)
+              for (t1, t2) in D.points]
+
   njobs = length(all_jobs)
 
-  @sync @distributed for job = 1:length(all_jobs)
-    (ind_n, (c_index, cdag_index)), (n1, t1), (n2, t2) = all_jobs[job]
+  # Transfer computed matrix elements to process id 1
+  # as (t1, t2, c_n, cdag_n, value) tuples
+  gf_element_channel = RemoteChannel(
+    ()->Channel{Tuple{TimeGridPoint,TimeGridPoint,Int,Int,ComplexF64}}(njobs),
+    1
+  )
+
+  @sync @distributed for job = 1:njobs
+    (c_n, c_index), (cdag_n, cdag_index), (t1, t2) = all_jobs[job]
```

Comment:
can we factor this inner code to compute a single element of the Green's function out into its own function?